core/server/master/src/main/java/alluxio/underfs/UfsStatusCache.java

```diff
@@ -0,0 +1,275 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.underfs;
+
+import alluxio.AlluxioURI;
+import alluxio.collections.ConcurrentHashSet;
+import alluxio.exception.InvalidPathException;
+import alluxio.master.file.meta.MountTable;
+import alluxio.resource.CloseableResource;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.RejectedExecutionException;
+
+/**
+ * This class is a cache from an Alluxio namespace URI ({@link AlluxioURI}, i.e. /path/to/inode) to
+ * UFS statuses.
+ *
+ * It also allows associating a path with child inodes, so that the statuses for a specific path can
+ * be searched for later.
+ */
+public class UfsStatusCache {
+  private static final Logger LOG = LoggerFactory.getLogger(UfsStatusCache.class);
+
+  private final ConcurrentHashMap<AlluxioURI, UfsStatus> mStatuses;
+  private final ConcurrentHashMap<AlluxioURI, Future<Collection<UfsStatus>>> mActivePrefetchJobs;
+  private final ConcurrentHashMap<AlluxioURI, Collection<UfsStatus>> mChildren;
+  private final ExecutorService mPrefetchExecutor;
+
+  /**
+   * Create a new instance of {@link UfsStatusCache}.
+   *
+   * @param prefetchExecutor the executor service used to prefetch statuses
+   */
+  public UfsStatusCache(@Nullable ExecutorService prefetchExecutor) {
+    mStatuses = new ConcurrentHashMap<>();
+    mChildren = new ConcurrentHashMap<>();
+    mActivePrefetchJobs = new ConcurrentHashMap<>();
+    mPrefetchExecutor = prefetchExecutor;
+  }
+
+  /**
+   * Add a new status to the cache.
+   *
+   * The last component of the path in the {@link AlluxioURI} must match the result of
+   * {@link UfsStatus#getName()}. This method overrides any status currently cached for the same
+   * URI.
+   *
+   * @param path the Alluxio path to key on
+   * @param status the ufs status to store
+   * @throws IllegalArgumentException if the status name doesn't match the final URI path component
+   */
+  public void addStatus(AlluxioURI path, UfsStatus status) {
+    UfsStatus prev = mStatuses.putIfAbsent(path, status);
```

Comment:
since it's a cache, I just ignore the previous value. We should never add the same status twice to the cache with the current code path, but if it does happen I think overwriting it is fine.

I will update this method to return the old status if there is one. No one in the current code will use the result though.

Reply:
updated